src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,219 @@
+      trait = makeKey(trait);
+      String formattedValue;
+      if (value instanceof Date) {
+        Date dateValue = (Date) value;
+        formattedValue = FIREBASE_FORMAT.format(dateValue);
+      } else {
+        formattedValue = String.valueOf(value);
+      }
+      firebaseAnalytics.setUserProperty(trait, formattedValue);
+      logger.verbose("firebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+    }
+  }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
+  @Override
+  public void track(TrackPayload track) {
+    super.track(track);
 
+    String event = track.event();
+    String eventName = makeKey(event);
+    Properties properties = track.properties();
+    Bundle formattedProperties = formatProperties(properties);
+    firebaseAnalytics.logEvent(eventName, formattedProperties);
+    logger.verbose("firebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
   }
 
-  @Override public void screen(ScreenPayload screen) {
-    super.screen(screen);
-//    mFirebaseAnalytics.setCurrentScreen(this, screen.name(), null /* class override */);
+  private Bundle formatProperties(Properties properties) {
+    Bundle bundle = new Bundle();
+    if ((properties.revenue() != 0 || properties.total() != 0)
+        && isNullOrEmpty(properties.currency())) {
+      bundle.putString(Param.CURRENCY, "USD");
+    }
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      Object value = entry.getValue();
+      String property = entry.getKey();
+      property = makeKey(property);
+      if (value instanceof Integer) {
+        int intValue = (int) value;
+        bundle.putInt(property, intValue);
+        logger.verbose("bundle.putInt(%s, %s);", property, intValue);
+        continue;
+      }
+      if (value instanceof Double) {
+        double doubleValue = (double) value;
+        bundle.putDouble(property, doubleValue);
+        logger.verbose("bundle.putDouble(%s, %s);", property, doubleValue);
+        continue;
+      }
+      if (value instanceof Long) {
+        long longValue = (long) value;
+        bundle.putLong(property, longValue);
+        logger.verbose("bundle.putLong(%s, %s);", property, longValue);
+        continue;
+      }
+      if (value instanceof String) {
+        String stringValue = String.valueOf(value);
+        bundle.putString(property, stringValue);
+        logger.verbose("bundle.putString(%s, %s);", property, stringValue);
+        continue;
+      }
+      if (value instanceof Date) {
+        Date dateValue = (Date) value;
+        String formattedDate = FIREBASE_FORMAT.format(dateValue);
+        bundle.putString(property, formattedDate);
+        logger.verbose("bundle.putString(%s, %s);", property, formattedDate);
+      }
+    }
+    return bundle;
   }
 
-  @Override public void track(TrackPayload track) {
-    super.track(track);
-
+  private String makeKey(String key) {
+    if (EVENT_MAPPER.containsKey(key)) {
```

Comment:
I don't think this shouldn't live here. The trim/replace work is the only generic component. 

Right now this has some bugs. e.g. you call `trait = makeKey(trait);`. If somebody sends a trait `name`, it will be mapped to `Param.ITEM_NAME`. I don't think that's what we want.


Hence the implementation should only be:

```
private static String makeKey(String key) {
return key.trim().replaceAll(" ", "_").substring(0, Math.min(key.length(), 40));
}
```

and then the mapping check should be outside this function usage (using track as an example) should be

```java
String event = track.event();
if (EVENT_MAPPER.containsKey(event)) {
  event = EVENT_MAPPER.get(key);
}
event = makeKey(event);
...
```